### PR TITLE
Enable Build/CI with Mono

### PR DESCRIPTION
**Pull request by dsyme**
This enables a build of this repo using Mono. Initially I've been testing using Travis - we should move to .Jenkins, perhaps as part of this PR.

* Use FSharp.Compiler.Tools nuget package as LKG (like fsharp/fsharp)
* Update ``init-tools.cmd`` and ``init-tools.sh`` to be the latest from the .NET Core repos.  
* There are a bunch of simplifications and extra checking in FSharpSource,targets
* Some of the key signing logic has Mono-compat implementations
* Removed use of tuple structs from the compiler code (this was done when initially updating the LKG. I eventually updated the LKG to one that actually supports tuple structs, but the uses of the feature aren't needed in the compiler code, so I'll remove them in any case)
* Remove the awful CustomCopyLocal and ``subst.exe``.  We can work out how to implement a copy-and-substitute at a later point.
* Project files on Mono need to use

```
   <FSharpSourcesRoot>$(MSBuildProjectDirectory)\..\..</FSharpSourcesRoot>
```
instead of
```
    <FSharpSourcesRoot>..\..</FSharpSourcesRoot>
```
* CoreCLR bits are not built on Mono though the PR includes some preliminary changes in that direction.

We do some things differently on Linux/Mono:
* UseMicroBuild  is off
* UseSourceLink is off
* UseGatherBinaries is off
* AddVsSdkAttributesToSomeCoreComponents is off
* The define CROSS_PLATFORM_COMPILER is set when building for Mono to workaround a couple of limitations in Mono.  The code is written in a way so that it is valid to set it on Windows too, but we don't CI that.

One odd de-capitalization change was needed in ``init-tools.sh`` to use ``microsoft.dotnet.buildtools`` instead of ``Microsoft.DotNet.BuildTools`` since the Linux package gets the former name

### CODE REVIEW

**Comment by dsyme** [comment]
This is ready for review.  An example green Travis run is here: https://travis-ci.org/dsyme/visualfsharp/jobs/173129120

**Comment by dsyme** [comment]
All green on Mono 4.6.1 and Mono latest https://travis-ci.org/dsyme/visualfsharp/builds/173131056

Just have to get all green and have the changes reviewed here please

File: DEVGUIDE.md [hunk 1]
```diff
@@ -23,15 +23,31 @@ The primary technical documents for the F# compiler code are
 
 ## Quick Start: Build, Test, Develop
 
-You can build the F# compiler for .NET Framework as follows:
+### F# Compiler (Linux)
+
+Currently you can do on Linux a bootstrap of the Mono version of the compiler.  Full testig is not enabled,
```
Comment (smoothdeveloper) [inline DEVGUIDE.md]: typo "full testing"

File: DEVGUIDE.md [hunk 2]
```diff
@@ -23,15 +23,31 @@ The primary technical documents for the F# compiler code are
...
+On Windows you can build the F# compiler for .NET Framework as follows:
 
     build.cmd
 
 This is the same as
 
     build.cmd net40
 
-There are various qualifiers:
+There are various qualifiers you can do to do more:
```
Comment (smoothdeveloper) [inline DEVGUIDE.md]: "you can use to do"

File: DEVGUIDE.md [hunk 3]
```diff
@@ -23,15 +23,31 @@ The primary technical documents for the F# compiler code are
 
 ## Quick Start: Build, Test, Develop
 
-You can build the F# compiler for .NET Framework as follows:
+### F# Compiler (Linux)
```
Comment (smoothdeveloper) [inline DEVGUIDE.md]: It mentions Linux but in before_install.sh the apt-get command are, I assume, ubuntu specific, maybe mention "tested with Ubuntu"?

Comment (dsyme) [inline DEVGUIDE.md]: thanks

File: DEVGUIDE.md (hunk 2 above)
Comment (dsyme) [inline DEVGUIDE.md]: thanks, fixed

**Comment by forki** [comment]
This makes me happy.

**Comment by 7sharp9** [comment]
How do you install into the mono prefix you are using?

**Comment by dsyme** [comment]
@7sharp9 You can't install at the moment, just bootstrap LKG --> Proto --> debug/release

This doesn't enable a strong dev/test path nor negate the reasons for http://github.com/fsharp/fsharp, but brings this a bit closer.

**Comment by dsyme** [comment]
I extend this PR with a change to make ``fssrgen`` an F# script invoked via ``Exec`` rather than a nugget package tool.  This is dramatically simpler.

**Comment by vasily-kirichenko** [comment]
It builds successfully on Ubuntu, but I cannot find any artifacts like fsc.exe / fsi.exe.

**Comment by dsyme** [comment]
@vasily-kirichenko We were missing a ``BUILD_NET40=1``, see [this change to build.sh](https://github.com/Microsoft/visualfsharp/pull/1703/commits/bfcd9a5c0f04eb47b9d99f893dd50adf6ea779d8). We can somehow start to unify build.sh and build.cmd logic now to avoid this sort of thing.

Artifacts should be under Debug

**Comment by dsyme** [comment]
We're no longer green on Travis:
```
				Executing: /home/travis/build/dsyme/visualfsharp/src/fsharp/FSharp.Compiler-proto/../../../packages/FSharp.Compiler.Tools.4.0.1.19/tools/fsi.exe --exec /home/travis/build/dsyme/visualfsharp/src/scripts/fssrgen.fsx "/home/travis/build/dsyme/visualfsharp/src/fsharp/FSComp.txt" "obj/Proto/net40/FSComp.fs"  "obj/Proto/net40/FSComp.resx" 
				args = [|"/home/travis/build/dsyme/visualfsharp/src/scripts/fssrgen.fsx";
				  "/home/travis/build/dsyme/visualfsharp/src/fsharp/FSComp.txt";
				  "obj/Proto/net40/FSComp.fs"; "obj/Proto/net40/FSComp.resx"|]
				/home/travis/build/dsyme/visualfsharp/src/fsharp/FSComp.txt(0): error : An exception occurred when processing
```

**Comment by dsyme** [comment]
Travis is green again https://travis-ci.org/dsyme/visualfsharp/branches

**Comment by forki** [comment]
+1,324 −21,315 to make it work on mono.

giphy

**Comment by dsyme** [comment]
@forki Mostly the LKG going.  But using FSharp.Compiler.Tools has also uncovered a few problems, especially on Mono (executable bits not set, using F# Interactive scripts on Mono causes codegen failures for some variations on ``fssrgen.fsx``). I'll try to capture those as issues.

**Comment by dsyme** [comment]
@forki See https://github.com/fsharp/fsharp/issues/639 and https://github.com/fsharp/fsharp/issues/638

**Pull request merged**